src/NUnitFramework/framework/Internal/Execution/TestWorker.cs

```diff
@@ -135,19 +137,28 @@ public void Start()
             _workerThread.Start();
         }
 
+        private object cancelLock = new object();
+
         /// <summary>
         /// Stop the thread, either immediately or after finishing the current WorkItem
         /// </summary>
-        public void Cancel()
+        /// <param name="force">true if the thread should be aborted, false if it should allow the currently running test to complete</param>
+        public void Cancel(bool force)
         {
-            _running = false;
+            if (force)
+                _running = false;
 
-#if NETCF
-            if (_workerThread != null && !_workerThread.Join(0))
-#else
-            if (_workerThread != null && _workerThread.IsAlive)
-#endif
-                ThreadUtility.Kill(_workerThread);
+            lock (cancelLock)
+                if (_workerThread != null && _currentWorkItem != null)
+                {
+                    {
+                        _currentWorkItem.Context.ExecutionStatus = force ? TestExecutionStatus.AbortRequested : TestExecutionStatus.StopRequested;
```

Comment:
OK, here is the answer to my earlier question. This is where the topmost work item has its context set.

I think this would be clearer if it all happened in the Cancel method. Also, it does not appear that you handle the top-level context in SimpleWorkItemDispatcher.